File: ios/TodoNative/Services/CompanionActions.swift
import Foundation

@MainActor
enum CompanionActions {
    static let allowedKinds = ["add_task", "complete_task", "breakdown"]

    struct SuggestedAction: Identifiable, Equatable {
        let id: UUID
        let label: String
        let kind: String
        let payload: [String: String]

        init(label: String, kind: String, payload: [String: String]) {
            self.id = UUID()
            self.label = label
            self.kind = kind
            self.payload = payload
        }
    }

    static func parse(_ reply: String, maxActions: Int = 2) -> (text: String, actions: [SuggestedAction]) {
        let text = reply.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let start = text.firstIndex(of: "{"), let end = text.lastIndex(of: "}"), start < end else {
            return (text, [])
        }
        let head = String(text[..<start]).trimmingCharacters(in: .whitespacesAndNewlines)
        let tail = String(text[text.index(after: end)...]).trimmingCharacters(in: .whitespacesAndNewlines)
        guard let jsonData = String(text[start...end]).data(using: .utf8),
              let json = (try? JSONSerialization.jsonObject(with: jsonData)) as? [String: Any],
              let rawActions = json["actions"] as? [[String: Any]] else {
            return (text, [])
        }
        var actions: [SuggestedAction] = []
        for raw in rawActions {
            guard actions.count < maxActions else { break }
            guard let kind = raw["action"] as? String, allowedKinds.contains(kind) else { continue }
            let payload = (raw["payload"] as? [String: Any])?.compactMapValues { $0 as? String } ?? [:]
            let label = raw["label"] as? String ?? kind
            actions.append(SuggestedAction(label: label, kind: kind, payload: payload))
        }
        let clean = [head, tail].filter { !$0.isEmpty }.joined(separator: "\n")
        return (clean.isEmpty ? text : clean, actions)
    }
}

File: companion-actions.js
const ALLOWED_KINDS = ['add_task', 'complete_task', 'breakdown'];

export function parseBuddyActions(reply, maxActions = 2) {
    const text = String(reply || '').trim();
    if (!text) return { text: '', actions: [] };

    const firstBrace = text.indexOf('{');
    const lastBrace = text.lastIndexOf('}');
    if (firstBrace < 0 || lastBrace <= firstBrace) {
        return { text, actions: [] };
    }

    const head = text.slice(0, firstBrace).trim();
    const tail = text.slice(lastBrace + 1).trim();
    let json;
    try {
        json = JSON.parse(text.slice(firstBrace, lastBrace + 1));
    } catch (_error) {
        return { text, actions: [] };
    }

    const rawActions = Array.isArray(json && json.actions) ? json.actions : [];
    const actions = [];
    for (const raw of rawActions) {
        if (actions.length >= maxActions) break;
        if (!raw || typeof raw !== 'object') continue;
        const kind = raw.action;
        if (!ALLOWED_KINDS.includes(kind)) continue;
        const payload = raw.payload && typeof raw.payload === 'object'
            ? Object.fromEntries(Object.entries(raw.payload).filter(([, v]) => typeof v === 'string'))
            : {};
        const label = typeof raw.label === 'string' && raw.label ? raw.label : kind;
        actions.push({ action: kind, payload, label });
    }

    const clean = [head, tail].filter(Boolean).join('\n');
    return { text: clean || text, actions };
}

if (typeof window !== 'undefined') {
    window.parseBuddyActions = parseBuddyActions;
}